drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,512 @@
+	}
+
+	err = adc_emul_validate_buffer_size(dev, sequence);
+	if (err) {
+		LOG_ERR("buffer size too small");
+		return err;
+	}
+
+	data->res_mask = BIT_MASK(sequence->resolution);
+	data->buffer = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_emul_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	struct adc_emul_data *data = dev->data;
+	int err;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	err = adc_emul_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static int adc_emul_read(const struct device *dev,
+			 const struct adc_sequence *sequence)
+{
+	return adc_emul_read_async(dev, sequence, NULL);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_emul_data *data = CONTAINER_OF(ctx, struct adc_emul_data,
+						  ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+/**
+ * @brief Main function of thread which is used to collect samples from
+ *        emulated ADC. When adc_context_start_sampling give semaphore,
+ *        for each requested channel value function is called. Returned
+ *        mV value is converted to output using reference voltage, gain
+ *        and requested resolution.
+ *
+ * @param data Internal data of ADC emulator
+ *
+ * @return This thread should not end
+ */
+static void adc_emul_acquisition_thread(struct adc_emul_data *data)
+{
+	struct adc_emul_channel_cfg *channel_cfg;
+	adc_emul_res_t result = 0;
+	uint32_t input_mV = 0;
+	uint8_t channel;
+	uint32_t ref_v;
+	uint64_t temp; /* Temporary 64 bit value prevent overflows */
+	int err;
+
+	while (true) {
+		k_sem_take(&data->sem, K_FOREVER);
+
+		err = 0;
+
+		while (data->channels) {
+			channel = find_lsb_set(data->channels) - 1;
+
+			LOG_DBG("reading channel %d", channel);
+
+			channel_cfg = &data->channels_cfg[channel];
+			err = channel_cfg->func(data->dev, channel,
+						channel_cfg->param, &input_mV);
+			if (err) {
+				LOG_ERR("failed to read channel %d (err %d)",
+					channel, err);
+				adc_context_complete(&data->ctx, err);
+				break;
+			}
+
+			ref_v = adc_emul_get_ref_voltage(data,
+							 channel_cfg->ref);
+			err = adc_gain_invert(channel_cfg->gain, &ref_v);
+			if (ref_v == 0 || err) {
+				LOG_ERR("failed to get ref voltage (ch %d)",
+					channel);
+				adc_context_complete(&data->ctx, -EINVAL);
+				break;
+			}
+
+			temp = (uint64_t)input_mV * data->res_mask / ref_v;
+
+			if (temp > data->res_mask)
+				temp = data->res_mask;
```

Comment:
I wonder if all the loop code above could go in a function, like read_value() ? This functoin is quite long.

Reply:
I extracted this code to `adc_emul_get_chan_value()` function.